feat: App should detect the user payed for the movie already

mwaa opened the pull request:
Please remember the [Contributing Guidelines](https://github.com/leapdao/meta/blob/master/CONTRIBUTION.md) :heart:

## What does this PR do?
<!-- Concise description of what this PR achieves, including any context. -->
- Post movie tokenID in message property of send request
- Stores transaction hash and other receipt properties in local storage
- Check localstorage first before wallet popup for payment request
- Notification when movie is already paid for

## Recommendations for testing

1. npm run start-local
2. Navigate to http://localhost:1234
3. Select a movie trailler and make payment
4. Refresh page and select trailer in previous step
5. Receive Thanks notification instead of wallet popup


## Links to relevant issues or information

Closes #19

Clearwood commented on packages/client/src/index.ts:
```diff
@@ -158,6 +176,26 @@ class WhiteRabbitClient {
+  getFromLedger(tokenId: string)
+  {
+    let ledger = JSON.parse(localStorage.getItem('ledger') || '[]');
+    for (let index = 0; index < ledger.length; index++) {
```
the performance would increase if the data structure would not be an array but key value pairs with the key being the token ID and there would be a direct check for the token ID

Clearwood commented on packages/client/src/index.ts:
```diff
@@ -158,6 +176,26 @@ class WhiteRabbitClient {
+      if (receipt.message === tokenId) {
+        return receipt;
+      }
+    }
```
The expected behavior is here that if no results are found in local storage the transactions of the user should be pulled and checked through as well as saved to local storage.

troggy replied:
Burner keeps its key in the same localStorage, so the situation when the key signed a transfer and the payment record is not in localStorage can happen only when the user imports his private key from other WR instance. I don't know whether it is realistic use case cause we are not expecting blockchain-savvy users. 

cc @johannbarbie

Clearwood replied:
The question is whether this is a use case we want to acknowledge, being able to synchronize purchases across devices in that way or across browsers.

troggy replied:
> being able to synchronize purchases across devices in that way or across browsers

I talking about "across streaming sites opened on a same device in a same browser"

Clearwood replied:
sure :) you are right that it is probably overkill for our application to support the import of old transactions.

troggy commented:
Thanks for your contribution! :👏

In addition to @Clearwood's comments I'd add the following concerns here:

1.  Payment information should be shared across all the streaming sites
It is one-to-many relationship between WR app (aka wallet) and streaming sites. There can be many streaming sites connected to WR system and some of these sites may be streaming the same movie. The proposed solution doesn't support that — if there are two sites S1 and S2 streaming the same movie X and I pay for it on site S1, site S2 will ask me to pay for movie X again.

Thus we want to always source information about payment from the WR app, we should not calculate it on the client side. I'd see that as additional message interaction between client and wallet which happens when `requestPayment` is called.

2. We want the user to clearly see the separation between streaming site and WhiteRabbit. And we care that WhiteRabbit brand is visible. Thus appreciation message should be framed in WhiteRabbit context (for instance in the same popup we use for payment)

Clearwood commented:
I wanted to add to @troggy comments: Local storage can not be accessed through different domains which means it has to be always accessed through the iFrame as only then it will always be accessed through the same context to be available accross different streaming sites.

troggy commented:
@mwaa I can rebase your PR on top of the updated `develop` branch, if you like. Are there any uncommitted changes in your working copy?

mwaa commented:
@Clearwood @troggy addressed comments.
Payment check is done in wallet and reusing `PaidMoviePage`. 
Still using local storage as burner wallet is dependent on it. 
When we allow wallet export/import then we would need to fetch user transaction history from 
blockchain

troggy commented on packages/burner-plugin/src/ui/MoviePage.tsx:
```diff
@@ -70,6 +74,17 @@ const MoviePage: React.FC<MoviePageContext> = ({
+      const event = events[index];
+      if (event.to === movieOwner && event.message === tokenId) 
+      {
+        history.push(`/paid/${tokenId}/${event.tx}`);
```
nit pick: in case this routine is slow (e.g. because of the large tx history) user will see payment screen, at least flickering for a moment.

Clearwood commented on packages/burner-plugin/src/ui/MoviePage.tsx:
```diff
@@ -70,6 +74,17 @@ const MoviePage: React.FC<MoviePageContext> = ({
   });
 
+  useEffect(() => {
+    const events = actions.getHistoryEvents({ account });
```
what if rather than filtering by the from address you would filter by the movieOwner address, that way we might improve performance on this :) otherwise this looks nice

troggy replied:
@Clearwood your logic doesn't check that payment was sent by `account`, it rather checks that "someone paid for the movie"

troggy replied:
IMO `actions.getHistoryEvents({ account })` is better because there won't be a lot of txs for a single user account (thus smaller history to scan).  Tx history size for viewer account is proportional to the appetites and payment capacity of the viewer thus can grow only reasonably. Tx history size for movie owner is proportional to movie popularity and thus can be unreasonably big

mwaa replied:
Made change

Clearwood replied:
@troggy the history is only saved locally, thus checking the local tx action history for a movie owner will be smaller than checking for all transactions of the account.

Clearwood replied:
I looked through burner core (https://github.com/austintgriffith/burner-core/blob/develop/core/src/History.js) and thought a lot about this proposed change, the only possibility could be that with a previously imported private key a movie was paid on that machine on that browser in which case I would argue it is fine to not ask the user for a payment again.

troggy replied:
shit.. it doesn't make sense to check history at all because of that :) It is way more easier to just make a payment record in localStorage on a wallet side and check it

troggy replied:
this bounty took longer then it should.. let's wrap it up. I suggest we just store `tokenId` for paid movie in a localStorage somewhere in`PaidMovie.tsx` near the `postMessage`. And in MoviePage here we just check the `localStorage` for the `tokenId` in paid movies list

troggy replied:
@mwaa @Clearwood what do you think?

troggy replied:
I'm also fine with leaving it as it is — it is working as it should, it is readable and reasonably fast

Clearwood commented on packages/burner-plugin/src/ui/MoviePage.tsx:
```diff
@@ -70,6 +74,17 @@ const MoviePage: React.FC<MoviePageContext> = ({
+    const events = actions.getHistoryEvents({ account });
+    for (let index = 0; index < events.length; index++) {
+      const event = events[index];
+      if (event.to === movieOwner && event.message === tokenId) 
```
```suggestion
      if (event.message === tokenId) 
```

troggy commented on packages/burner-plugin/src/ui/MoviePage.tsx:
```diff
@@ -78,14 +78,13 @@ const MoviePage: React.FC<MoviePageContext> = ({
-          break;
-        }      
+      const events = actions.getHistoryEvents({ account });
+      const paidEvent = events.find((event) => {
```
agreed that `.find` is more readable and is fast enough for the expected tx history size for viewer account

Clearwood replied:
I agree, I think this is great :). Could you filter by destination address in your call to `actions.getHistoryEvents({ movieOwner })` as recommended by me yesterday?

mwaa replied:
As discussed using `account` in order to ensure payment was sent by `account`.
Check comments from Kosta

troggy replied:
@Clearwood as I understand [burner-core](https://github.com/austintgriffith/burner-core/blob/develop/core/src/History.js#L58), the `actions.getHistoryEvents({ account: movieOwner })` call will return all the transactions sent to and from `movieOwner` account. The counterparty in these transactions may be any other account, but we only want to find a tx from _our_ account (which is in `account` variable).

troggy replied:
oh, well.. @Clearwood just pointed out that getHistoryEvents contains only the txs for the viewers account (because it comes from localStorage, never from the network). So, yes — we can use `actions.getHistoryEvents({ account: movieOwner })` as filtering
